### PR TITLE
Solving #2117 Conv2d not working in MPC

### PR DESCRIPTION
**Pull request by Jasopaum**
first commit: implementation without bias, padding, dilation

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 1]
```diff
@@ -476,6 +512,130 @@ def argmax(tensor, **kwargs):
...
+                else:
+                    res = res + bias
+
+            # ... And reshape it back to an image
+            res = (
+                res.permute(0, 2, 1)
+                .view(batch_size, nb_channels_out, nb_rows_out, nb_cols_out)
+                .contiguous()
+            )
+            return res
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py]: You can do a bit simpler with the .child stuff here:
```
                res = torch.cat(res, dim=2)
            else:
                res = im_reshaped.matmul(weight_reshaped)

            # Add a bias if needed
            if bias is not None:
                res += bias

            # ... And reshape it back to an image
            res = (
                res.permute(0, 2, 1)
                .view(batch_size, nb_channels_out, nb_rows_out, nb_cols_out)
                .contiguous()
            )
            return res.child
```

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 2]
```diff
@@ -505,6 +665,16 @@ def relu(tensor_shares):
...
+                def pad(input, pad, mode="constant", value=0):
+                    padded_shares = {}
+                    shares_dict = input.child.child if input.is_wrapper else input.child
+                    for location, shares in shares_dict.items():
+                        padded_shares[location] = torch.nn.functional.pad(shares, pad, mode, value)
+
+                    return AdditiveSharingTensor(padded_shares)
+
+                module.pad = pad
+
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py]: You should use the decorator `@overloaded.function` which helps you to have a simpler code:
```
                @overloaded.function
                def pad(input_shares, pad, mode="constant", value=0):
                    padded_shares = {}
                    for location, shares in input_shares.items():
                        padded_shares[location] = torch.nn.functional.pad(shares, pad, mode, value)

                    return padded_shares

                module.pad = pad
```

**Comment by LaRiffle** [comment]
Ah I said it is a really good PR!
From the tests I observed something is not clear - you can (should) never share additively floats (well in theory because you succeeded here ahah). Float should be converted to fixedprecision tensors
This way, you actually need to define conv2d only in fixed_precision.py not in additive_shared.py (I just tried it it works). This avoids duplicate code.
Also, the code in conv2d is really tricky to understand, you might want to add a few comments to explain how you transform your data to do the convolution in a single matrix multiplication, or link any useful reference.

**Comment by Jasopaum** [comment]
Thanks for the review! I will modify the code and create a new PR.

For the additively shared floats, they actually were not floats but int tensors that's why it worked when I shared them. Do you still think we should remove conv2d from AdditivelySharingTensor? Maybe the users would like to share int tensors without using the fix_prec intermediary (I don't know if this makes sense).
